File: src/main/java/me/lokka30/levelledmobs/listeners/EntityTransformListener.java
/*
 * Copyright (c) 2020-2021  lokka30. Use of this source code is governed by the GNU AGPL v3.0 license that can be found in the LICENSE.md file.
 */

package me.lokka30.levelledmobs.listeners;

import java.util.Collections;
import java.util.HashSet;
import java.util.List;

import me.lokka30.levelledmobs.LevelledMobs;
import me.lokka30.levelledmobs.managers.DebugManager;
import me.lokka30.levelledmobs.misc.AdditionalLevelInformation;
import me.lokka30.levelledmobs.misc.DebugType;
import me.lokka30.levelledmobs.misc.LevellableState;
import me.lokka30.levelledmobs.wrappers.LivingEntityWrapper;
import me.lokka30.levelledmobs.rules.LevelledMobSpawnReason;
import org.bukkit.entity.Entity;
import org.bukkit.entity.LivingEntity;
import org.bukkit.event.EventHandler;
import org.bukkit.event.EventPriority;
import org.bukkit.event.Listener;
import org.bukkit.event.entity.EntitySpawnEvent;
import org.bukkit.event.entity.EntityTransformEvent;
import org.jetbrains.annotations.NotNull;

/**
 * Listens for when a mob transforms so the applicable rules can be applied
 *
 * @author stumper66
 * @version 2.4.0
 */
public class EntityTransformListener implements Listener {

    private final LevelledMobs main;

    public EntityTransformListener(final LevelledMobs main) {
        this.main = main;
    }

    @EventHandler(ignoreCancelled = true, priority = EventPriority.MONITOR)
    public void onTransform(@NotNull final EntityTransformEvent event) {
        // is the original entity a living entity
        if (!(event.getEntity() instanceof LivingEntity)) {
            DebugManager.log(DebugType.ENTITY_MISC, event.getEntity(), false, () ->
                    event.getEntity().getType().name() +
                ": entity was &bnot&7 an instance of LivingEntity");
            return;
        }

        // is the original entity levelled
        if (!main.levelManager.isLevelled((LivingEntity) event.getEntity())) {
            DebugManager.log(DebugType.ENTITY_MISC, event.getEntity(), false, () ->
                    event.getEntity().getName() + ": original entity was &bnot&7 levelled");
            if (event.getTransformReason() == EntityTransformEvent.TransformReason.SPLIT)
                checkForSlimeSplit((LivingEntity) event.getEntity(), event.getTransformedEntities());
            return;
        }

        final LivingEntityWrapper lmEntity = LivingEntityWrapper.getInstance(
            (LivingEntity) event.getEntity(), main);

        boolean useInheritance = false;
        int level = 1;

        if (main.rulesManager.getRuleMobLevelInheritance(lmEntity)) {
            useInheritance = true;
            level = lmEntity.getMobLevel();
        }

        for (final Entity transformedEntity : event.getTransformedEntities()) {
            if (!(transformedEntity instanceof LivingEntity)) {
                DebugManager.log(DebugType.ENTITY_MISC, event.getEntity(), false, () ->
                        event.getEntity().getType().name()
                        + ": entity was&b not&7 an instance of LivingEntity (loop)");
                continue;
            }

            final LivingEntityWrapper transformedLmEntity = LivingEntityWrapper.getInstance(
                (LivingEntity) transformedEntity, main);
            final LevellableState levelledState = main.levelInterface.getLevellableState(
                transformedLmEntity);
            if (levelledState != LevellableState.ALLOWED) {
                DebugManager.log(DebugType.ENTITY_MISC, event.getEntity(), false, () ->
                        transformedEntity.getType().name()
                        + ": transformed entity was &bnot&7 levellable, reason: &b"
                        + levelledState);
                main.levelManager.updateNametagWithDelay(transformedLmEntity);
                transformedLmEntity.free();
                continue;
            }

            DebugManager.log(DebugType.ENTITY_MISC, event.getEntity(), true, () ->
                    transformedEntity.getType().name() + ": entity was transformed");

            if (useInheritance) {
                if (lmEntity.getSpawnReason() == LevelledMobSpawnReason.LM_SPAWNER) {
                    transformedLmEntity.setSpawnReason(LevelledMobSpawnReason.SPAWNER);
                }

                main.levelInterface.applyLevelToMob(
                    transformedLmEntity,
                    level,
                    false,
                    false,
                    new HashSet<>(Collections.singletonList(
                        AdditionalLevelInformation.FROM_TRANSFORM_LISTENER))
                );
            } else {
                main.levelManager.entitySpawnListener.preprocessMob(transformedLmEntity,
                    new EntitySpawnEvent(transformedEntity));
            }

            main.levelManager.updateNametagWithDelay(lmEntity);
            transformedLmEntity.free();
        }

        lmEntity.free();
    }

    private void checkForSlimeSplit(final @NotNull LivingEntity livingEntity, final @NotNull List<Entity> transformedEntities){
        final LivingEntityWrapper parent = LivingEntityWrapper.getInstance(livingEntity, main);
        if (parent.getSpawnReason() == LevelledMobSpawnReason.DEFAULT ||
                parent.getSpawnReason() == LevelledMobSpawnReason.SLIME_SPLIT){
            parent.free();
            return;
        }

        for (final Entity transformedEntity : transformedEntities) {
            if (!(transformedEntity instanceof final LivingEntity le)) continue;

            final LivingEntityWrapper lew = LivingEntityWrapper.getInstance(le, main);
            lew.setSpawnReason(parent.getSpawnReason());
            lew.free();
        }

        parent.free();
    }
}

File: levelledmobs-plugin/src/main/kotlin/io/github/arcaneplugins/levelledmobs/listeners/EntityTransformListener.kt
package io.github.arcaneplugins.levelledmobs.listeners

import io.github.arcaneplugins.levelledmobs.LevelledMobs
import io.github.arcaneplugins.levelledmobs.debug.DebugManager
import io.github.arcaneplugins.levelledmobs.result.AdditionalLevelInformation
import io.github.arcaneplugins.levelledmobs.debug.DebugType
import io.github.arcaneplugins.levelledmobs.enums.InternalSpawnReason
import io.github.arcaneplugins.levelledmobs.enums.LevellableState
import io.github.arcaneplugins.levelledmobs.util.Log
import io.github.arcaneplugins.levelledmobs.wrappers.LivingEntityWrapper
import java.util.UUID
import org.bukkit.Bukkit
import org.bukkit.entity.Entity
import org.bukkit.entity.LivingEntity
import org.bukkit.entity.Player
import org.bukkit.event.EventPriority
import org.bukkit.event.HandlerList
import org.bukkit.event.Listener
import org.bukkit.event.entity.CreatureSpawnEvent
import org.bukkit.event.entity.EntitySpawnEvent
import org.bukkit.event.entity.EntityTransformEvent

/**
 * Listens for when a mob transforms so the applicable rules can be applied
 *
 * @author stumper66
 * @version 2.4.0
 */
class EntityTransformListener : Listener {
    val damageMappings = mutableMapOf<UUID, Player>()
    private var lastPriority: EventPriority? = null
    private val settingName = "entity-transform-event"

    companion object {
        @JvmStatic
        lateinit var instance: EntityTransformListener
            private set
    }

    init {
        instance = this
    }

    fun load(){
        val priority = LevelledMobs.instance.mainCompanion.getEventPriority(settingName, EventPriority.MONITOR)
        if (lastPriority != null){
            if (priority == lastPriority) return

            HandlerList.unregisterAll(this)
            Log.inf("Changing event priority for $settingName from $lastPriority to $priority")
        }

        Bukkit.getPluginManager().registerEvent(
            EntityTransformEvent::class.java,
            this,
            priority,
            { _, event -> if (event is EntityTransformEvent) onTransform(event) },
            LevelledMobs.instance,
            false
        )
        lastPriority = priority
    }

    private fun onTransform(event: EntityTransformEvent) {
        // is the original entity a living entity
        if (event.entity !is LivingEntity) {
            DebugManager.log(DebugType.ENTITY_MISC, event.entity, false) {
                "entity was &bnot&7 an instance of LivingEntity"
            }
            return
        }

        val main = LevelledMobs.instance
        // is the original entity levelled
        if (!main.levelManager.isLevelled(event.entity as LivingEntity)) {
            DebugManager.log(DebugType.ENTITY_MISC, event.entity, false) {
                "original entity was &bnot&7 levelled"
            }
            if (event.transformReason == EntityTransformEvent.TransformReason.SPLIT)
                checkForSlimeSplit(event.entity as LivingEntity,event.transformedEntities)
            return
        }

        val lmEntity = LivingEntityWrapper.getInstance(event.entity as LivingEntity)

        var useInheritance = false
        var level = 1

        if (main.rulesManager.getRuleMobLevelInheritance(lmEntity)) {
            useInheritance = true
            level = lmEntity.getMobLevel
        }

        for (transformedEntity in event.transformedEntities) {
            if (transformedEntity !is LivingEntity) {
                DebugManager.log(DebugType.ENTITY_MISC, event.entity, false) {
                    "entity was&b not&7 an instance of LivingEntity (loop)"
                }
                continue
            }

            val transformedLmEntity = LivingEntityWrapper.getInstance(transformedEntity)

            val levelledState: LevellableState = main.levelInterface.getLevellableState(
                transformedLmEntity
            )
            if (levelledState != LevellableState.ALLOWED) {
                DebugManager.log(DebugType.ENTITY_MISC, event.entity, false) {
                    "transformed entity was &bnot&7 levellable, reason: &b$levelledState"
                }
                main.levelManager.updateNametagWithDelay(transformedLmEntity)
                transformedLmEntity.free()
                continue
            }

            DebugManager.log(
                DebugType.ENTITY_MISC,
                transformedEntity,
                true
            ) { "entity was transformed" }

            if (useInheritance) {
                val internalSpawnReason = lmEntity.spawnReason.getInternalSpawnReason(lmEntity)
                if (internalSpawnReason == InternalSpawnReason.LM_SPAWNER)
                    transformedLmEntity.spawnReason.setMinecraftSpawnReason(lmEntity, CreatureSpawnEvent.SpawnReason.SPAWNER)

                main.levelInterface.applyLevelToMob(
                    transformedLmEntity,
                    level,
                    isSummoned = false,
                    bypassLimits = false,
                    additionalLevelInformation = mutableSetOf(AdditionalLevelInformation.FROM_TRANSFORM_LISTENER)
                )
            } else {
                main.levelManager.entitySpawnListener.processMob(
                    transformedLmEntity,
                    EntitySpawnEvent(transformedEntity)
                )
            }

            main.levelManager.updateNametagWithDelay(lmEntity)
            transformedLmEntity.free()
        }

        lmEntity.free()
    }

    private fun checkForSlimeSplit(livingEntity: LivingEntity, transformedEntities: List<Entity>) {
        val parent = LivingEntityWrapper.getInstance(livingEntity)
        val minecraftSpawnReason = parent.spawnReason.getMinecraftSpawnReason(parent)
        if (minecraftSpawnReason == CreatureSpawnEvent.SpawnReason.DEFAULT ||
            minecraftSpawnReason == CreatureSpawnEvent.SpawnReason.SLIME_SPLIT
        ) {
            parent.free()
            return
        }

        for (transformedEntity in transformedEntities) {
            if (transformedEntity !is LivingEntity) continue

            val lew = LivingEntityWrapper.getInstance(transformedEntity)
            lew.spawnReason.setLMSpawnReason(parent)
            lew.free()
        }

        parent.free()
    }
}